core/src/config/extrinsic_params.rs

```diff
@@ -28,15 +28,82 @@ pub trait ExtrinsicParams<T: Config>: ExtrinsicParamsEncoder + Sized + Send + 's
 /// defines how to encode the "additional" and "extra" params. Both functions
 /// are optional and will encode nothing by default.
 pub trait ExtrinsicParamsEncoder: 'static {
-    /// This is expected to SCALE encode the "signed extra" parameters
-    /// to some buffer that has been provided. These are the parameters
-    /// which are sent along with the transaction, as well as taken into
-    /// account when signing the transaction.
-    fn encode_extra_to(&self, _v: &mut Vec<u8>) {}
-
-    /// This is expected to SCALE encode the "additional" parameters
-    /// to some buffer that has been provided. These parameters are _not_
-    /// sent along with the transaction, but are taken into account when
+    /// This is expected to SCALE encode the transaction extension data to some
+    /// buffer that has been provided. This data is attached to the transaction
+    /// and also (by default) attached to the signer payload which is signed to
+    /// provide a signature for the transaction.
+    ///
+    /// If [`ExtrinsicParamsEncoder::encode_signer_payload_value_to`] is implemented,
+    /// then that will be used instead when generating a signer payload. Useful for
+    /// eg the `VerifySignature` extension, which is send with the transaction but
+    /// is not a part of the signer payload.
+    fn encode_value_to(&self, _v: &mut Vec<u8>) {}
+
+    /// See [`ExtrinsicParamsEncoder::encode_value_to`]. This defaults to calling that
+    /// method, but if implemented will dictate what is encoded to the signer payload.
+    fn encode_signer_payload_value_to(&self, v: &mut Vec<u8>) {
+        self.encode_value_to(v);
+    }
+
+    /// This is expected to SCALE encode the "implicit" (formally "additional")
+    /// parameters to some buffer that has been provided. These parameters are
+    /// _not_ sent along with the transaction, but are taken into account when
     /// signing it, meaning the client and node must agree on their values.
-    fn encode_additional_to(&self, _v: &mut Vec<u8>) {}
+    fn encode_implicit_to(&self, _v: &mut Vec<u8>) {}
+
+    /// Set the signature. This happens after we have constructed the extrinsic params,
+    /// and so is defined here rather than on the params, below. We need to use `&dyn Any`
+    /// to keep this trait object safe, but can downcast in the impls.
+    fn inject_signature(&mut self, _account_id: &dyn Any, _signature: &dyn Any) {}
```

Comment:
aye, I see the downcast/upcast annoyance now

Reply:
Yup exactly; the impl has access to `T: Config` and the types come from a client with the same config, so all should work, but yeah let's make the warnings way clearer here!